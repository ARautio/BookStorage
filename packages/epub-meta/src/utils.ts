import { MetadataProps } from './index';

const parseAuthor = (authors: any) => {
  const authorString = authors[0]._;
  return authorString.split(',').map((author: string) => author.trim());
};

const parseDate = (dates: any[]) => {
  const date = dates[0] ? dates[0] : undefined;
  if (typeof date === 'object') {
    return new Date(date._);
  }
  return date !== undefined ? new Date(date) : undefined;
};

const parseISBN = (identifiers: any) => {
  const isbnIdent = identifiers.find(
    (ident: any) => ident['$']['opf:scheme'] === 'ISBN'
  );
  return isbnIdent?._;
};

const parseCoverPath = (meta: any, manifest: any) => {
  const coverMetaName = meta.find(
    (metaItem: any) => metaItem['$']['name'] === 'cover'
  );
  const coverItem = manifest.find(
    (item: any) => item['$'].id === coverMetaName['$']['content']
  );
  if (coverItem !== undefined) {
    return coverItem['$'].href;
  }
  return undefined;
};

export const parseMeta = (metadata: any): MetadataProps => {
  const data = metadata?.package?.metadata[0];
  const manifest = metadata?.package?.manifest[0].item;
  return {
    title: data['dc:title'],
    author: parseAuthor(data['dc:creator']),
    date: parseDate(data['dc:date']),
    isbn: parseISBN(data['dc:identifier']),
    coverPath: parseCoverPath(data.meta, manifest),
  };
};

export const parseCover = (epub: any) => {
  return epub;
};
