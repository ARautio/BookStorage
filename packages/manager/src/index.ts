import express from "express";
import sqlite from "sqlite3";
import cors from "cors";
import * as http from "http";
import * as ws from "ws";

import { books } from "./endpoints/books";
import { opds } from "./endpoints/opds";
import { BookRepository } from "./repository/bookRepository";
import { EPubRepository } from "./repository/epubRepository";
import { SettingsRepository } from "./repository/settingsRepository";
import Book from "./models/book";
import BookFile from "./models/bookfile";

const app = express();
app.use(cors());

const PORT = 8000;
const server = http.createServer(app);
const wss = new ws.Server({ noServer: true });

const db = new sqlite.Database(":memory:");

const settingsRepository = new SettingsRepository(db);
settingsRepository.initiate();

const ePubRepository = new EPubRepository(Book, BookFile);
const bookRepository = new BookRepository(db);
bookRepository.initiate();

books({ app, wss }, bookRepository, ePubRepository, settingsRepository);
opds({ app, wss }, bookRepository);

app.get("/", (req, res) => res.send("BookStorage started"));

app.use("/assets/books/", express.static("../../books"));
app.use("/assets/covers/", express.static("../../covers"));

wss.on("connection", (socket) => {
  socket.send(JSON.stringify({ action: "CONNECTION" }));
});

wss.on("close", (ws: any) => {
  ws.send("connection close");
});

server.on("upgrade", function (request, socket, head) {
  wss.handleUpgrade(request, socket, head, function (ws) {
    wss.emit("connection", ws, request);
  });
});

server.listen(PORT, () => {
  console.log(`⚡️[server]: Server is running at https://localhost:${PORT}`);
});
