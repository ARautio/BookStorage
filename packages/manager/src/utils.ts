import sqlite3 from "sqlite3";

export const getPath = (basePath: string) => (path: string): string =>
  `${basePath}${path}`;

export const getFromDB = (db: sqlite3.Database, query: string) => {
  return new Promise((resolve, reject) => {
    db.all(query, (error, rows) => {
      if (error) reject(error);
      else resolve(rows);
    });
  });
};

export const getSingleFromDB = (db: sqlite3.Database, query: string) => {
  return new Promise((resolve, reject) => {
    db.get(query, (error, row) => {
      if (error) reject(error);
      else resolve(row);
    });
  });
};

export const runFromDB = (db: sqlite3.Database, query: string) => {
  return new Promise((resolve, reject) => {
    db.run(query, (error) => {
      if (error) reject(error);
      else resolve(null);
    });
  });
};
