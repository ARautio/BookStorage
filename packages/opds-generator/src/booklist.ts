interface IBookList {
  baseUrl: String;
  type: String;
  kind: String;
  title: String;
  updated: Date;
  books: IBook[];
}

export const root = ({
  baseUrl,
  type,
  kind,
  title,
  updated,
  books,
}: IBookList) => {
  return `
  <?xml version="1.0" encoding="UTF-8"?>
<feed xmlns="http://www.w3.org/2005/Atom"
      xmlns:dc="http://purl.org/dc/terms/"
      xmlns:opds="http://opds-spec.org/2010/catalog">
  <id>urn:uuid:433a5d6a-0b8c-4933-af65-4ca4f02763eb</id>

  <link rel="self"    
        href="${baseUrl}/${type}.xml"
        type="application/atom+xml;profile=opds-catalog;kind=${kind}"/>
  <link rel="start"  
        href="${baseUrl}/root.xml"
        type="application/atom+xml;profile=opds-catalog;kind=navigation"/>

  <title>${title}</title>
  <updated>${updated}</updated>
  <author>
    <name>Spec Writer</name>
    <uri>http://opds-spec.org</uri>
  </author> 

  ${books.map(item => book(item)).join('')}
</feed>
`.trim();
};

interface IBook {
  title: String;
  uuid: String;
  authors: IAuthor[];
  updated: Date;
  isbn: String;
  language: String;
  issued: Date;
  publisher: String;
  coverfilename: String;
  description: Date;
  coverFilename: String;
  files: IBookFile[];
  settings: Settings;
}

interface Settings {
  filePath: string;
  coverPath: string;
}

interface IAuthor {
  name: String;
}

interface IBookFile {
  filename: String;
  filetype: String;
}

const book = ({
  title,
  uuid,
  authors,
  updated,
  isbn,
  language,
  publisher,
  coverfilename,
  issued,
  description,
  files,
  settings: { coverPath, filePath },
}: IBook) => {
  return `
      <entry>
      <title>${title}</title>
      <id>urn:uuid:${uuid}</id>
      ${authors
        .map(
          author => `
      <author>
        <name>${author.name}</name>
        <uri>http://opds-spec.org/authors/21285</uri>
      </author>
      `
        )
        .join('')}
      <updated>${updated}</updated>
      <dc:identifier>urn:isbn:${isbn}</dc:identifier>
      <dc:publisher>${publisher}</dc:publisher>
      <dc:language>${language}</dc:language>
      <dc:issued>${issued}</dc:issued>
      <content type="text">${description}</content>
      <link rel="http://opds-spec.org/image"    
            href="${coverPath}/${coverfilename}"
            type="image/jpeg"/>
        
      ${files.map(
        file => `
        <link rel="http://opds-spec.org/acquisition/buy"
              href="${filePath}/${file.filename}"
              type="${file.filetype}">
        </link>
        `
      )}
   </entry>
      `;
};
