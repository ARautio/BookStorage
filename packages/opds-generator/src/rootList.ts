interface IRoot {
  baseDir: String;
  updated: Date;
}

export const root = ({ baseDir = '/opds-catalog', updated }: IRoot) => {
  const rootList: [IRootEntry] = [
    {
      title: 'Popular books',
      type: 'http://opds-spec.org/sort/popular',
      link: `${baseDir}/popular.xml`,
      updated: new Date(),
      description: 'Yeah popular',
    },
  ];

  return `
  <?xml version="1.0" encoding="UTF-8"?>
  <feed xmlns="http://www.w3.org/2005/Atom">
    <id>urn:uuid:2853dacf-ed79-42f5-8e8a-a7bb3d1ae6a2</id>
    <link rel="self"  
          href="${baseDir}/root.xml"
          type="application/atom+xml;profile=opds-catalog;kind=navigation"/>
    <link rel="start"
          href="${baseDir}/root.xml"
          type="application/atom+xml;profile=opds-catalog;kind=navigation"/>
    <title>OPDS Catalog Root Example</title>
    <updated>${updated}</updated>
    <author>
      <name>Spec Writer</name>
      <uri>http://opds-spec.org</uri>
    </author>
    ${rootList.join('\n')}
  </feed>
  `;
};

interface IRootEntry {
  title: String;
  type: String;
  link: String;
  updated: Date;
  description: String;
}

export const rootEntry = ({
  title,
  type,
  link,
  updated,
  description,
}: IRootEntry) => {
  return `
  <entry>
    <title>${title}</title>
    <link rel=${type}
          href=${link}
          type="application/atom+xml;profile=opds-catalog;kind=acquisition"/>
    <updated>${updated}</updated>
    <id>urn:uuid:d49e8018-a0e0-499e-9423-7c175fa0c56e</id>
    <content type="text">${description}</content>
  </entry>
  `;
};
