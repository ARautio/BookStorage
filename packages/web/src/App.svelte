<script lang="ts">
  import { Router, Route } from "svelte-routing";
  import { QueryClient, QueryClientProvider } from "@sveltestack/svelte-query";
  import { onMount } from "svelte";
  import socket from "./api/socket";

  export let url = ""; //This property is necessary declare to avoid ignore the Router

  import Layout from "containers/Layout/Layout.svelte";
  import BookList from "containers/BookList/BookList.svelte";

  const queryClient = new QueryClient();
  onMount(() => {
    socket.subscribe((data) => {
      console.log(data);
    });
  });
</script>

<QueryClientProvider client={queryClient}>
  <Router {url}>
    <Layout>
      <Route path="/" component={BookList} />
      <Route path="books" component={BookList} />
    </Layout>
  </Router>
</QueryClientProvider>
