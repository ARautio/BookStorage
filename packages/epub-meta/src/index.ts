import zipFile from 'node-stream-zip';
import { parseStringPromise } from 'xml2js';
import mime from 'mime-types';

import * as utils from './utils';
import { MetadataProps } from '../types/index';

class EPub {
  filepath: string;
  metadata: MetadataProps | undefined;
  cover: any;

  constructor(filepath: string) {
    this.filepath = filepath;
    this.metadata = undefined;
    this.cover = null;
  }

  async load() {
    const zip = new zipFile.async({ file: this.filepath });
    const file = await zip.entryData('OEBPS/content.opf');
    const data = await parseStringPromise(file.toString());
    this.metadata = utils.parseMeta(data);
  }

  async getCover() {
    if (this.metadata === undefined) {
      throw Error('Load file data first');
    }
    if (this.metadata.coverPath === undefined) {
      throw Error('This book doesnt have cover');
    }
    const zip = new zipFile.async({ file: this.filepath });
    const file = await zip.entryData(`OEBPS/${this.metadata.coverPath}`);
    return { file, mimetype: mime.lookup(this.metadata.coverPath) };
  }
}

export default EPub;
