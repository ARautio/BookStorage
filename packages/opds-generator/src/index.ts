import * as bookList from './booklist';
import * as rootList from './rootList';

export const book = bookList;
export const root = rootList;
