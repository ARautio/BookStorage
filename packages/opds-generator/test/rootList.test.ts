import * as rootList from '../src/rootList';
import xmlParser from 'fast-xml-parser';

describe('rootList', () => {
  describe('root', () => {
    it('should form valid xml', () => {
      expect(
        xmlParser.validate(
          rootList.root({
            baseDir: '/',
            updated: new Date('2020-01-01'),
            options: [
              {
                title: 'Popular',
                type: 'entry',
                link: '/books',
                updated: new Date('2020-01-02'),
                description: 'Popular books',
              },
            ],
          })
        )
      ).toEqual(true);
    });

    it('should generate base item', () => {
      expect(
        rootList.root({
          baseDir: '/',
          updated: new Date('2020-01-01'),
          options: [
            {
              title: 'Popular',
              type: 'entry',
              link: '/books',
              updated: new Date('2020-01-02'),
              description: 'Popular books',
            },
          ],
        })
      ).toMatchSnapshot();
    });
  });
  describe('rootEntry', () => {
    it('should generate an entry', () => {
      expect(
        rootList.rootEntry({
          title: 'Test Book',
          type: 'entry',
          link: '/books',
          updated: new Date('2020-01-01'),
          description: 'Trest',
        })
      ).toMatchSnapshot();
    });
  });
});
