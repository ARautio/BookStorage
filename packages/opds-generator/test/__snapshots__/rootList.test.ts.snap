// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rootList root should generate base item 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
  <feed xmlns=\\"http://www.w3.org/2005/Atom\\">
    <id>urn:uuid:2853dacf-ed79-42f5-8e8a-a7bb3d1ae6a2</id>
    <link rel=\\"self\\"  
          href=\\"//root.xml\\"
          type=\\"application/atom+xml;profile=opds-catalog;kind=navigation\\"/>
    <link rel=\\"start\\"
          href=\\"//root.xml\\"
          type=\\"application/atom+xml;profile=opds-catalog;kind=navigation\\"/>
    <title>OPDS Catalog Root Example</title>
    <updated>Wed Jan 01 2020 02:00:00 GMT+0200 (Eastern European Standard Time)</updated>
    <author>
      <name>Spec Writer</name>
      <uri>http://opds-spec.org</uri>
    </author>
    
  <entry>
    <title>Popular</title>
    <link rel=\\"entry\\"
          href=\\"/books\\"
          type=\\"application/atom+xml;profile=opds-catalog;kind=acquisition\\"/>
    <updated>Thu Jan 02 2020 02:00:00 GMT+0200 (Eastern European Standard Time)</updated>
    <id>urn:uuid:d49e8018-a0e0-499e-9423-7c175fa0c56e</id>
    <content type=\\"text\\">Popular books</content>
  </entry>
  
  </feed>"
`;

exports[`rootList rootEntry should generate an entry 1`] = `
"
  <entry>
    <title>Test Book</title>
    <link rel=\\"entry\\"
          href=\\"/books\\"
          type=\\"application/atom+xml;profile=opds-catalog;kind=acquisition\\"/>
    <updated>Wed Jan 01 2020 02:00:00 GMT+0200 (Eastern European Standard Time)</updated>
    <id>urn:uuid:d49e8018-a0e0-499e-9423-7c175fa0c56e</id>
    <content type=\\"text\\">Trest</content>
  </entry>
  "
`;
